internal/storage/fs/snapshot.go

```diff
@@ -1,14 +1,618 @@
 package fs
 
 import (
+	"context"
+	"encoding/json"
 	"errors"
+	"fmt"
 	"io"
+	iofs "io/fs"
+	"path"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/gofrs/uuid"
+	ferrors "go.flipt.io/flipt/errors"
+	"go.flipt.io/flipt/internal/ext"
+	"go.flipt.io/flipt/internal/storage"
+	"go.flipt.io/flipt/rpc/flipt"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	indexFile = ".flipt.yml"
+	defaultNs = "default"
 )
 
-type storeSnapshot struct{}
+var ErrNotImplemented = errors.New("not implemented")
+
+// FliptIndex represents the structure of a well-known file ".flipt.yml"
+// at the root of an FS.
+type FliptIndex struct {
+	Version    string   `yaml:"version,omitempty"`
+	Inclusions []string `yaml:"inclusions,omitempty"`
+	Exclusions []string `yaml:"exclusions,omitempty"`
+}
+
+// storeSnapshot contains the structures necessary for serving
+// flag state to a client.
+type storeSnapshot struct {
+	ns        map[string]*namespace
+	evalDists map[string][]*storage.EvaluationDistribution
+}
+
+type namespace struct {
+	resource  *flipt.Namespace
+	flags     map[string]*flipt.Flag
+	segments  map[string]*flipt.Segment
+	rules     map[string]*flipt.Rule
+	evalRules map[string][]*storage.EvaluationRule
+}
+
+func newNamespace(key string) *namespace {
+	return &namespace{
+		resource:  &flipt.Namespace{Key: key, Name: strings.Title(key)},
+		flags:     map[string]*flipt.Flag{},
+		segments:  map[string]*flipt.Segment{},
+		rules:     map[string]*flipt.Rule{},
+		evalRules: map[string][]*storage.EvaluationRule{},
+	}
+}
+
+// snapshotFromReaders constructs a storeSnapshot from the provided
+// slice of io.Reader.
+func snapshotFromReaders(sources ...io.Reader) (*storeSnapshot, error) {
+	s := storeSnapshot{
+		ns: map[string]*namespace{
+			defaultNs: {
+				resource: &flipt.Namespace{
+					Key:  defaultNs,
+					Name: "Default",
+				},
+			},
+		},
+		evalDists: map[string][]*storage.EvaluationDistribution{},
+	}
+
+	for _, reader := range sources {
+		doc := new(ext.Document)
+
+		if err := yaml.NewDecoder(reader).Decode(doc); err != nil {
+			return nil, err
+		}
+
+		if err := s.addDoc(doc); err != nil {
+			return nil, err
+		}
+
+	}
+	return &s, nil
+}
+
+func listStateFiles(logger *zap.Logger, source iofs.FS) ([]string, error) {
+	// This is the default variable + value for the FliptIndex. It will preserve its value if
+	// a .flipt.yml can not be read for whatever reason.
+	idx := FliptIndex{
+		Version: "1.0",
+		Inclusions: []string{
+			"**/features.yml", "**/features.yaml", "**/*.features.yml", "**/*.features.yaml",
+		},
+	}
+
+	// Read index file
+	inFile, err := source.Open(indexFile)
+	if err == nil {
+		if derr := yaml.NewDecoder(inFile).Decode(&idx); derr != nil {
+			return nil, derr
```

Comment:
should we be wrapping these returned errors in any of these unexported funcs? or are we wrapping the error from the caller?

Reply:
Good q. I think that the caller will just log out the error in this case. I do not think it will need to assert against the error or anything like that from the caller.